### PR TITLE
targetcli: 2.1.fb49 -> 2.1.51

### PR DESCRIPTION
**Pull request by markuskowa**
##### Motivation for this change
Regular update. Now the man page is also installed.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: pkgs/os-specific/linux/targetcli/default.nix [hunk 1]
```diff
@@ -2,17 +2,22 @@
...
 
   src = fetchFromGitHub {
     owner = "open-iscsi";
     repo = "${pname}-fb";
     rev = "v${version}";
-    sha256 = "093dmwc5g6yz4cdgpbfszmc97i7nd286w4x447dvg22hvwvjwqhh";
+    sha256 = "0xrvby63i39rvi778bnvnxacghaix63q72vzxdc3i87ji1ki58hc";
   };
 
   propagatedBuildInputs = with python.pkgs; [ configshell rtslib ];
```
Comment (ajs124) [inline pkgs/os-specific/linux/targetcli/default.nix]: configshell and rtslib seem to also have gotten new releases, so they should probably be bumped as well.

Comment (markuskowa) [reply pkgs/os-specific/linux/targetcli/default.nix]: Good catch, thanks.

**Comment by ajs124** [comment]
With the dependencies updated, we arrive at the age old question: did you test this with an actual iSCSI setup?

**Comment by markuskowa** [comment]
I did a local test with the the open-iscsi initiator and the targetcli/fileio backend on my laptop (before I updated the libraries), which worked fine. A great thing to have would be a NixOS test, that covers a few basic scenarios.

**Comment by markuskowa** [comment]
@GrahamcOfBorg test iscsi

**Comment by markuskowa** [comment]
The test works just fine when I build it locally. Any ideas why it fails with ofBorg?

**Comment by ajs124** [comment]
Seems to work for me locally, as well. @grahamc do you have any insights?

File: nixos/tests/iscsi.nix [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    # Setup target
+    $server->succeed("targetcli iscsi/ create ${targetName}");
+    $server->succeed("targetcli iscsi/${targetName}/tpg1/acls/ create ${initiatorName}");
+
+    $server->succeed("targetcli iscsi/${targetName}/tpg1/luns create storage_object=/backstores/fileio/image1");
+    $server->succeed("targetcli iscsi/${targetName}/tpg1/luns create storage_object=/backstores/block/image2");
+
+    # Check client login
+    $client->waitForUnit("multi-user.target");
+    $client->succeed("iscsistart -i ${initiatorName} -t ${targetName} -g 1 -a server");
```
Comment (Mic92) [inline nixos/tests/iscsi.nix]: Maybe a race condition?

```suggestion
    $client->waitUntilSucceeds("iscsistart -i ${initiatorName} -t ${targetName} -g 1 -a server");
```

**Comment by JohnAZoidberg** [comment]
This is awesome!

I've been trying out the iSCSI infrastructure on NixOS recently and I'm planning to create a test and maybe module for `tgt`, as it allows declarative configuration. So I'd suggest to rename this test to `targetcli`.

**Comment by markuskowa** [comment]
@JohnAZoidberg I was also looking at `tgt`. However, it seems that this package is quite old and the recommended version is now `targetcli`, which is said to be more performant (uses the kernel space iscsi target). There might be a chance to use declarative configuration with `targetcli`: the tool stores and reads from a json file. Maybe that could be integrated in nix module somehow?

File: nixos/tests/iscsi.nix (hunk 2 above)
Comment (markuskowa) [reply nixos/tests/iscsi.nix]: Thanks that seems to work.

**Review by JohnAZoidberg — APPROVED** [review]
Even though `tgt` is still maintained, it looks like `targetcli` is more or less a replacement:

- https://lwn.net/Articles/424004/
- https://groveronline.com/2012/11/tgtd-lio-in-rhel-7/

Yes, it does look like we can use the JSON config to declaratively configure it. I wasn't aware of that before. I'll try that.
Alright, I'm for having `targetcli` as NixOS' default iSCSI toolchain.

File: nixos/tests/iscsi.nix (hunk 2 above)
Comment (markuskowa) [reply nixos/tests/iscsi.nix]: Sorry, I got something wrong here. Still fails with ofBorg.

Comment (Mic92) [reply nixos/tests/iscsi.nix]: Is the command necessary after all or what does:

```
iscsistart: initiator reported error (15 - session exists)
```

mean?

Comment (markuskowa) [reply nixos/tests/iscsi.nix]: `iscsistart` connects manually to the server (target). It is required to create new iSCSI provided block devices on the client ("/dev/sda" and "/dev/sdb" in this case). The message `session exists` means that the connect worked only half way at the first try (it somehow connected to target but still failed).

**Comment by JohnAZoidberg** [comment]
This should be good, even works, when rebased on top of the latest master.

I'm working on the declarative service module but the JSON config seems to be too complicated to generate, completely, ourselves.

**Comment by markuskowa** [comment]
I think we we can leave the test in Perl for now and convert it later when new Python test frame work is more mature.

**Comment by JohnAZoidberg** [comment]
It sure looks like the Python framework is on-par and we're going all in: https://github.com/NixOS/nixpkgs/issues/72828

I'm quite stumped by this sudden change without a proper RFC or announcement :/

**Comment by jonringer** [comment]
I guess there was some discussion at nixcon, and there seemed to be consensus

**Comment by markuskowa** [comment]
Yes, the switch to python was quite quick but it I think it is a good thing to move away from perl.

The test is now in python and works fine (only GrahamcOfBorg seems to still have problems with it).

File: pkgs/os-specific/linux/targetcli/default.nix [hunk 3]
```diff
@@ -2,17 +2,22 @@
...
 
+  postInstall = ''
+    mkdir -p $out/share/man/man8
+    cp targetcli.8 $out/share/man/man8/
```
Comment (JohnAZoidberg) [inline pkgs/os-specific/linux/targetcli/default.nix]:
```suggestion
    install -D targetcli.8 -t $out/share/man/man8/
```
And remove the line above.

**Review by JohnAZoidberg — APPROVED** [review]

**Comment by markuskowa** [comment]
@jonringer should we merge it now?

**Comment by jonringer** [comment]
when trying to run the tests  locally, i got:

```
client: must succeed: iscsistart -i iqn.2004-01.org.nixos:initiator -t iqn.2004-01.org.nixos:target -g 1 -a server
client # [   84.446956] Loading iSCSI transport class v2.0-870.
client # iscsistart: version 2.0-878
client # [   85.448936] iscsi: registered transport (tcp)
client # iscsistart: initiator reported error (15 - session exists)
client # [   86.409030] scsi host2: iSCSI Initiator over TCP/IP
client # [   86.518122]  connection1:0: detected conn error (1011)
client: output: iscsistart: Logging into iqn.2004-01.org.nixos:target server:3260,1

error: command `iscsistart -i iqn.2004-01.org.nixos:initiator -t iqn.2004-01.org.nixos:target -g 1 -a server` failed (exit code 1)
cleaning up
```

**Comment by jonringer** [comment]
@GrahamcOfBorg test iscsi

**Comment by markuskowa** [comment]
That is strange I can run the test locally.

**Comment by ajs124** [comment]
> @jonringer should we merge it now?

Upstream seems to have release 2.1.51 in the meantime.

**Comment by markuskowa** [comment]
I did update all packages to the latest version. The test seems to be more stable now x86 but still is a bit shaky on ARM. Any suggestions?

**Comment by JohnAZoidberg** [comment]
Test fails on x64 nixos for me :(

```
client # [  256.637003] systemd[1]: Reached target Multi-User System.
client # [  256.669568] systemd[1]: Startup finished in 1min 10.998s (kernel) + 3min 5.582s (userspace) = 4min 16.581s.
client: waiting for success: ping -c 1 server
(0.58 seconds)
client: must succeed: iscsistart -i iqn.2004-01.org.nixos:initiator -t iqn.2004-01.org.nixos:target -g 1 -a server
client # [  259.585977] Loading iSCSI transport class v2.0-870.
client # iscsistart: version 2.0-878
client # [  260.640047] iscsi: registered transport (tcp)
client # iscsistart: initiator reported error (15 - session exists)
client # [  261.415787] scsi host2: iSCSI Initiator over TCP/IP
client # [  261.688013]  connection1:0: detected conn error (1011)
client: output: iscsistart: Logging into iqn.2004-01.org.nixos:target server:3260,1

error: command `iscsistart -i iqn.2004-01.org.nixos:initiator -t iqn.2004-01.org.nixos:target -g 1 -a server` failed (exit code 1)
```

**Comment by markuskowa** [comment]
@JohnAZoidberg I am almost about to give up on the test. It seems to be reliably unreliable. My guess is that `iscsistart` does not like when there are more than two network interfaces available (there are some mentions on google). I can see  a way to eliminate the problem.
Using `iscsistart` to test the functionality seems just not to be the right approach. Maybe the best way is to implement some basic NixOS modules for targetcli/iscsid first and and then base the test on iscsid.

**Comment by JohnAZoidberg** [comment]
> Maybe the best way is to implement some basic NixOS modules for targetcli/iscsid first and and then base the test on iscsid/iscsiadm.

Yes, that seems reasonable. I'm already working on that.

But for now I suggest we just merge the package updates and you could open another WIP PR for the test, I'd be interested to figure it out.

Are you using iSCSI on NixOS for anything other than testing?

**Comment by JohnAZoidberg** [comment]
According to [this](https://support.citrix.com/article/CTX217580) my error and the one on ofBorg for ARM means that the initiator is timing out, while connecting to the target.

**Comment by markuskowa** [comment]
@JohnAZoidberg I am not using iSCSI at the moment but I want to in the future (hopefully with iSCSI root at some point). 
Should we proceed like this: I will remove the the test from this PR. Since you are already working on a NixOS module the test will fit better into your PR. Feel free to just copy the test if you want to use it as a starting point.

**Comment by JohnAZoidberg** [comment]
Alright, sounds good, you can remove it.

Ohh, iSCSI root, that'll be interesting and probably doesn't work out of the box on NixOS right now.

**Comment by markuskowa** [comment]
OK, done

> Ohh, iSCSI root, that'll be interesting and probably doesn't work out of the box on NixOS right now.

Yeah, that will probably require some more work to make it run smoothly.

**Pull request merged**